protocols/bftcosi/bftcosi.go

```diff
@@ -0,0 +1,615 @@
+	// channel for challenge during the commit phase
+	challengeCommitChan chan challengeCommitChan
+	// channel for response
+	responseChan chan responseChan
+	// channel to notify when we are done
+	done chan bool
+	// channel to notify when the prepare round is finished
+	prepareFinishedChan chan bool
+	// channel used to wait for the verification of the block
+	verifyChan chan bool
+
+	// exceptions given during the rounds that is used in the signature
+	tempExceptions []cosi.Exception
+
+	// last block computed
+	lastBlock string
+	// temporary buffer of "prepare" commitments
+	tempPrepareCommit []*cosi.Commitment
+	tpcMut            sync.Mutex
+	// temporary buffer of "commit" commitments
+	tempCommitCommit []*cosi.Commitment
+	tccMut           sync.Mutex
+	// temporary buffer of "prepare" responses
+	tempPrepareResponse []*cosi.Response
+	tprMut              sync.Mutex
+	// temporary buffer of "commit" responses
+	tempCommitResponse []*cosi.Response
+	tcrMut             sync.Mutex
+
+	// refusal to sign for the commit phase or not. This flag is set during the
+	// Challenge of the commit phase and will be used during the response of the
+	// commit phase to put an exception or to sign.
+	signRefusal bool
+
+	// onDoneCallback is the callback that will be called at the end of the
+	// protocol when all nodes have finished. Either at the end of the response
+	// phase of the commit round or at the end of a view change.
+	onDoneCallback func()
+
+	// onSignatureDone is the callback that will be called when a signature has
+	// been generated ( at the end of the response phase of the commit round)
+	onSignatureDone func(*BFTSignature)
+
+	// verificationFun will be called
+	// during the (start/handle) challenge prepare phase of the protocol
+	verificationFun VerificationFunction
+
+	// bool set to true when the final signature is produced
+	doneSigning chan bool
+	// lock associated
+	doneLock sync.Mutex
+	// threshold for how much exception
+	threshold int
+	// done processing is used to stop the processing of the channels
+	doneProcessing chan bool
+
+	// finale signature that this BFTCoSi round has produced
+	finalSignature *BFTSignature
+}
+
+// NewBFTCoSiProtocol returns a new bftcosi struct
+func NewBFTCoSiProtocol(n *sda.Node, verify VerificationFunction) (*ProtocolBFTCoSi, error) {
+	// initialize the bftcosi node/protocol-instance
+	bft := &ProtocolBFTCoSi{
+		Node:             n,
+		suite:            n.Suite(),
+		prepare:          cosi.NewCosi(n.Suite(), n.Private()),
+		commit:           cosi.NewCosi(n.Suite(), n.Private()),
+		verifyChan:       make(chan bool),
+		doneProcessing:   make(chan bool, 2),
+		doneSigning:      make(chan bool, 1),
+		verificationFun:  verify,
+		aggregatedPublic: n.EntityList().Aggregate,
+		threshold:        int(2.0 * math.Ceil(float64(len(n.Tree().List()))/3.0)),
+	}
+
+	// register channels
+	if err := n.RegisterChannel(&bft.announceChan); err != nil {
+		return bft, err
```

Comment:
Shouldn't it be `return nil, err`? because the `bft` is not valid at this point, no?